src/app/utils/xml-utils.ts

```diff
@@ -36,3 +36,74 @@ export function replaceNotWordChar(textContent: string) {
 export function removeSpaces(textContent: string) {
   return textContent.replace(/\s/g, '');
 }
+
+/**
+ * It removes excessive spaces, any tabulation, new lines and non-word characters
+ */
+export function normalizeSpaces(textContent: string) {
+  return textContent.replace(/[\s]{2,}|\n|\t|\r/g, ' ').trimStart().trimEnd();
+}
+
+/**
+* Significant text sometimes is split inside two or more text evt-element inside the main one, especially when it contains new line characters.
+* This function returns a string with all the text elements chained
+*/
+export function chainFirstChildTexts(elem: XMLElement, evtTextComplexElements: string[], evtInnerTextElements: string[]): string {
+  if (elem === undefined) { return ''; };
+  const evtTextElements = {
+    '#text': 'nodeValue',
+    'p': 'textContent',
+  };
+  let out = '';
+  elem.childNodes.forEach((node) => (evtTextElements[node.nodeName] !== undefined) ? out += node[ evtTextElements[node.nodeName] ] : (
+    evtTextComplexElements.includes(node.nodeName) ? out += chainDeepTexts(node, evtInnerTextElements) : '' ))
+
+  return out;
+}
+
+/**
+ * Retrieve and chain textContent of all descendents
+ * @param elem ChildNode
+ * @returns out string
+ */
+export function chainDeepTexts(elem: ChildNode, evtInnerTextElements: string[]): string {
+  const textProperty = 'textContent';
+  let out = '';
+  elem.childNodes.forEach((x) => (evtInnerTextElements.includes(x.nodeName)) ? ((x[textProperty] !== null) ? out += x[textProperty] : '') : '')
+
+  return out;
+}
+
+/**
+* Retrieve external elements searching between provided elements types and filtering on attributes base
+* It searches all document for a bibl with the correct xml:id.
+* It would be faster if we knew the id or unique element to search in
+*/
+export function getExternalElements(elem: XMLElement, attrSourceNames: string[], attrTargetName: string, elTypes: string): XMLElement[] {
+  const sourceIDs = attrSourceNames.map((x) => elem.getAttribute(x));
+  const sourcesToFind = sourceIDs.filter((x) => x).map((x) => x.replace('#',''));
+
+  if (sourcesToFind.length === 0) {
+    return [];
+  }
+
+  return Array.from(elem.ownerDocument.querySelectorAll<XMLElement>(elTypes))
+    .filter((x) => sourcesToFind.includes(x.getAttribute(attrTargetName)))
+}
+
+/**
+ * If an element's attribute 'type' has one of the provided values then it is considered an analogue
+ */
+export function isAnalogue(elem: XMLElement, markerAttrs: string[]): boolean {
+  return (markerAttrs.includes(elem.getAttribute('type')));
+}
+
+/**
+ * If an element has one of the provided attributes then it is considered a source
+ */
+export function isSource(elem: XMLElement, attrs: string[]): boolean {
```

Comment:
This is a specific function for analogue. Why in xml-utils?

Reply:
Same as before, because it's used in more than one file for more than one elements. It seemed worse to copy-paste it in both of them or to create another utils-file just for those